File: rust/tests/gyroscope/l3g4200d_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::gyroscope::{L3g4200dFull, L3g4200dMinimal, FS_500_DPS, ODR_200_HZ, FIFO_STREAM};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x68);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut gyro = L3g4200dMinimal::new(dev, addr, false).expect("init L3G4200D");

    let (x, y, z) = gyro.angular_rate().unwrap();
    check_true!(x.abs() < 50.0, "angular_rate_x_range", passed, failed);
    check_true!(y.abs() < 50.0, "angular_rate_y_range", passed, failed);
    check_true!(z.abs() < 50.0, "angular_rate_z_range", passed, failed);

    drop(gyro);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut gyro_full = L3g4200dFull::new(dev, addr, false).expect("init L3G4200D Full");
    check_true!(gyro_full.who_am_i().unwrap_or(0) == 0xD3, "who_am_i", passed, failed);

    gyro_full.configure(ODR_200_HZ, 0, FS_500_DPS).unwrap();
    let (x2, _, _) = gyro_full.angular_rate().unwrap();
    check_true!(x2.abs() < 500.0, "configure_then_read", passed, failed);

    check_true!(gyro_full.status().unwrap_or(0xFF) <= 0xFF, "status_readable", passed, failed);
    let t = gyro_full.temperature().unwrap_or(0);
    check_true!(t >= -50 && t <= 100, "temperature_range", passed, failed);

    gyro_full.set_full_scale(2000).unwrap();
    let (x3, _, _) = gyro_full.angular_rate().unwrap();
    check_true!(x3.abs() < 2000.0, "set_full_scale_2000", passed, failed);

    let samples = gyro_full.fifo_samples().unwrap_or(0);
    check_true!(samples <= 31, "fifo_samples_in_range", passed, failed);

    gyro_full.enable_fifo(FIFO_STREAM, 10).unwrap();
    check_true!(true, "enable_fifo_no_throw", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/gyroscope/l3g4200d_test/main.go
//go:build linux && !tinygo

// L3G4200D hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the L3G4200D check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/gyroscope"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x68"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	chip, err := gyroscope.NewL3G4200DMinimal(conn, false)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new minimal:", err)
		os.Exit(2)
	}
	x, y, z, err := chip.AngularRate()
	check("angular_rate_x_range", err == nil && x >= -50.0 && x <= 50.0)
	check("angular_rate_y_range", err == nil && y >= -50.0 && y <= 50.0)
	check("angular_rate_z_range", err == nil && z >= -50.0 && z <= 50.0)

	conn2, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection full:", err)
		os.Exit(2)
	}
	defer conn2.Close()

	full, err := gyroscope.NewL3G4200DFull(conn2, false)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new full:", err)
		os.Exit(2)
	}
	cid, err := full.WHOAMI()
	check("who_am_i", err == nil && cid == 0xD3)

	if err := full.Configure(gyroscope.L3G4200DODR200Hz, 0, gyroscope.L3G4200DFS500DPS); err == nil {
		check("configure", true)
	} else {
		check("configure", false)
	}

	tmp, err := full.Temperature()
	check("temperature_range", err == nil && tmp >= -50 && tmp <= 100)

	st, err := full.Status()
	check("status_readable", err == nil && st <= 0xFF)

	if err := full.SetFullScale(gyroscope.L3G4200DFS2000DPS); err == nil {
		check("set_full_scale_2000", true)
	} else {
		check("set_full_scale_2000", false)
	}

	if err := full.EnableFIFO(gyroscope.L3G4200DFIFOStream, 10); err == nil {
		check("enable_fifo", true)
	} else {
		check("enable_fifo", false)
	}
	samples, err := full.FIFOSamples()
	check("fifo_samples_in_range", err == nil && samples <= 31)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
